src/go/serviceconfig/service_config_fetcher_test.go

```diff
@@ -15,34 +15,138 @@
 package serviceconfig
 
 import (
-	"strings"
+	"fmt"
+	"net/http"
+	"net/http/httptest"
 	"testing"
 	"time"
 
 	"github.com/GoogleCloudPlatform/esp-v2/src/go/options"
-	"github.com/GoogleCloudPlatform/esp-v2/src/go/util"
+	"github.com/golang/protobuf/proto"
+
+	confpb "google.golang.org/genproto/googleapis/api/serviceconfig"
 )
 
-func TestServiceConfigFetcherTimeout(t *testing.T) {
+func initServiceManagementForTestServbiceConfigFetcherFetchConfig(t *testing.T, serviceConfig *confpb.Service, serviceName string) *httptest.Server {
+	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if wantPath := fmt.Sprintf("/v1/services/%s/configs/%s", serviceName, serviceConfig.Id); r.URL.Path != wantPath {
+			t.Fatalf("want path: %s, get path: %s", wantPath, r.URL.Path)
+		}
+		resp, err := genFakeServiceConfig(serviceConfig)
+		if err != nil {
+			t.Fatalf("fail to generate servicecontrol report response: %v", err)
+		}
+		_, _ = w.Write(resp)
+	}))
+}
+
+func genFakeServiceConfig(serviceConfig *confpb.Service) ([]byte, error) {
+	return proto.Marshal(serviceConfig)
+}
+
+func TestServiceConfigFetcherFetchConfig(t *testing.T) {
+	serviceConfigId := "test-config-id"
+	serviceName := "service-name"
+	serviceConfig := &confpb.Service{
+		Name: "foo",
+		Id:   serviceConfigId,
+	}
+
+	serviceManagementServer := initServiceManagementForTestServbiceConfigFetcherFetchConfig(t, serviceConfig, serviceName)
 	opts := options.DefaultConfigGeneratorOptions()
-	timeout := 1 * time.Second
-	opts.HttpRequestTimeout = timeout
+	opts.ServiceManagementURL = serviceManagementServer.URL
 
-	serviceConfigFetcher, err := NewServiceConfigFetcher(nil, opts, "service-name")
+	scf, err := NewServiceConfigFetcher(&opts, serviceName, func() (string, time.Duration, error) { return "access-token", time.Duration(60), nil })
```

Comment:
Strongly recommend here (and all other places) to define the func as a named var in the line above and pass the var. It's really difficult to read on github otherwise.